confidential-data-hub/kms/src/plugins/aliyun/credential.rs

```diff
@@ -113,3 +147,71 @@ impl Credential {
         bytes
     }
 }
+
+// implement Credential::CredentialEcsRamRole related function
+impl Credential {
+    pub(crate) fn new_credential_ecs_ram_role(ecs_ram_role_name: &str) -> Result<Self> {
+        let credential = Credential::CredentialEcsRamRole {
+            ecs_ram_role_name: ecs_ram_role_name.to_string(),
+        };
+
+        Ok(credential)
+    }
+
+    pub(crate) async fn get_ecs_ram_role_session_credential(
+        &self,
+    ) -> Result<(String, String, String)> {
+        match self {
+            Credential::CredentialEcsRamRole { ecs_ram_role_name } => {
+                let request_url = format!(
+                    "http://100.100.100.200/latest/meta-data/ram/security-credentials/{}",
+                    ecs_ram_role_name
+                );
+
+                let response = reqwest::get(&request_url).await?;
+                let body = if response.status().is_success() {
+                    response.text().await?
+                } else {
+                    return Err(anyhow!(
+                        "Request session_credential failed with status: {}",
+                        response.status()
+                    ));
+                };
+                let body_json: Value = serde_json::from_str(&body)?;
+
+                let session_ak = body_json["AccessKeyId"].as_str().unwrap().to_string();
+                let session_sk = body_json["AccessKeySecret"].as_str().unwrap().to_string();
+                let token = body_json["SecurityToken"].as_str().unwrap().to_string();
+
+                Ok((session_ak, session_sk, token))
+            }
+            _ => Err(anyhow!(
+                "Require 'CredentialEcsRamRole' in 'get_ecs_ram_role_session_credential()'!"
+            )),
+        }
+    }
+
+    pub(crate) fn sign_with_credential_ecs_ram_role(
+        &self,
+        str_to_sign: &str,
+        secret: &str,
+    ) -> Result<String> {
+        match self {
+            Credential::CredentialEcsRamRole { .. } => {
+                let pkey = PKey::hmac(secret.as_bytes())?;
+                let mut signer = Signer::new(MessageDigest::sha1(), &pkey)?;
+                signer.update(str_to_sign.as_bytes())?;
+                let signature = signer.sign_to_vec()?;
+                Ok(STANDARD.encode(signature))
+            },
+            _=> Err(anyhow!("Require 'Credential::CredentialClientKey' in 'sign_with_credential_ecs_ram_role()'!"))
+        }
+    }
+
+    pub(crate) fn urlencode_openapi(&self, s: &String) -> String {
```

Comment:
```suggestion
pub(crate) fn urlencode_openapi(&self, s: &str) -> String {
```

When use a reference to a String, `&str` is preferred. 

BTW, is this function actually used?

Reply:
Yeah, it's used in both `build_params()` and `do_request()` functions.

Reply:
Modified.